File: app/src/main/java/com/example/ama/questapp/presentation/base/ViewStateFactory.java
package com.example.ama.questapp.presentation.base;

import com.example.ama.questapp.presentation.list.di.MainQuestListScope;

/***
 * Class with patterns for view states
 * @param <T> type of data that view uses
 */

@MainQuestListScope
public class ViewStateFactory<T> {

    public ViewState<T> empty() {
        return new ViewState<>(false, null, null);
    }

    public ViewState<T> loading() {
        return new ViewState<>(true, null, null);
    }

    public ViewState<T> hasData(T data) {
        return new ViewState<>(false, data, null);
    }

    public ViewState<T> error(Throwable error) {
        return new ViewState<>(false, null, error);
    }
}

File: app/src/main/java/com/example/ama/questapp/presentation/base/ViewStateFactory.kt
package com.example.ama.questapp.presentation.base

class ViewStateFactory<T> {

    fun empty() = ViewState<T>(false, null, null)

    fun loading() = ViewState<T>(true, null, null)

    fun hasData(data: T) = ViewState<T>(false, data, null)

    fun error(error: Throwable) = ViewState<T>(false, null, error)
}